File: back/models/game.go
package models

import (
	"challenge/services"
	"time"
)

type Game struct {
	ID          int          `json:"id" gorm:"primaryKey"`
	Name        string       `json:"name" gorm:"type:varchar(100)"`
	Description string       `json:"description" gorm:"type:varchar(255)"`
	Image       string       `json:"image" gorm:"type:varchar(255)"`
	Tags        []string     `json:"tags" gorm:"json"`
	Tournaments []Tournament `json:"tournaments" gorm:"foreignKey:GameID"`
	CreatedAt   time.Time    `json:"created_at"`
	UpdatedAt   time.Time    `json:"updated_at"`
	DeletedAt   *time.Time   `json:"deleted_at"`
}

type CreateGameDto struct {
	Name        string   `json:"name" validate:"required"`
	Description string   `json:"description" validate:"required"`
	Image       string   `json:"image" validate:"required"`
	Tags        []string `json:"tags" validate:"required"`
}

type UpdateGameDto struct {
	Name        string   `json:"name"`
	Description string   `json:"description"`
	Image       string   `json:"image"`
	Tags        []string `json:"tags"`
}

func FindAllGames(query services.QueryFilter) ([]Game, error) {
	var games []Game

	value := DB.Model(&Game{}).
		Where("deleted_at IS NULL").
		Offset(query.GetSkip()).
		Where(query.GetWhere()).
		Where(query.GetSearch()).
		Order(query.GetSort()).
		Preload("Tournaments")

	if query.GetLimit() != 0 {
		value.Limit(query.GetLimit())
	}

	err := value.Find(&games).Error

	return games, err
}

func CountGameByName(name string) (int64, error) {
	var count int64

	err := DB.Model(&Game{}).
		Where("deleted_at IS NULL").
		Where("name = ?", name).
		Count(&count).Error

	return count, err
}

func (g *Game) Save() error {
	return DB.Save(g).Error
}

func (g *Game) FindOneById(id int) error {
	return DB.First(g, id).Error
}

func (g *Game) FindOne(key string, value any) error {
	return DB.Where(key, value).
		Where("deleted_at IS NULL").
		First(g).Error
}

func (g *Game) Delete() error {
	del := time.Now()
	g.DeletedAt = &del
	g.Name = "[deleted]"
	g.Description = "[deleted]"

	return g.Save()
}

func ClearGames() error {
	return DB.Exec("DELETE FROM games").Error
}

func (g *Game) GetTournaments() ([]Tournament, error) {
	var tournaments []Tournament
	err := DB.Model(&g).Association("Tournaments").Find(&tournaments)
	return tournaments, err
}

File: front/lib/core/models/game.dart
class Game {
  final int id;
  final String name;
  final String description;
  final String imageUrl;
  final List<String> tags;
  final String createdAt;
  final String updatedAt;

  Game({
    required this.id,
    required this.name,
    required this.description,
    required this.imageUrl,
    required this.tags,
    required this.createdAt,
    required this.updatedAt,
  });

  Game copyWith({
    int? id,
    String? name,
    String? description,
    String? imageUrl,
    List<String>? tags,
    String? createdAt,
    String? updatedAt,
  }) {
    return Game(
      id: id ?? this.id,
      name: name ?? this.name,
      description: description ?? this.description,
      imageUrl: imageUrl ?? this.imageUrl,
      tags: tags ?? this.tags,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  factory Game.fromJson(Map<String, dynamic> json) {
    return Game(
      id: json['id'],
      name: json['name'],
      description: json['description'],
      imageUrl: json['image'],
      tags: List<String>.from(json['tags'] ?? []),
      createdAt: json['created_at'],
      updatedAt: json['updated_at'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'description': description,
      'image': imageUrl,
      'tags': tags,
      'created_at': createdAt,
      'updated_at': updatedAt,
    };
  }
}
